fs/spiffs/src/spiffs_vfs.c

```diff
@@ -408,7 +416,8 @@ static int spiffs_open(FAR struct file *filep, FAR const char *relpath,
 
   /* Check of the file object already exists */
 
-  ret = spiffs_find_objhdr_pgndx(fs, (FAR const uint8_t *)relpath, &pgndx);
+  ret = spiffs_find_objhdr_pgndx(fs, (FAR const uint8_t *)spiffs_relpath,
```

Comment:
can we enhance spiffs_find_objhdr_pgndx and spiffs_fobj_create to handle with or without the leading '/' correclty?

Reply:
if you mean depends on a config, we can. but i feel spiffs_vfs.c is a better place to handle nuttx-specific glitches like this.

if you mean a runtime detection, i feel it's overkill.

Reply:
The change add local variable and memcpy which increase the stack usage and reduce the performance. And every caller of spiffs_find_objhdr_pgndx/spiffs_fobj_create need to do the same thing repeatly. If we handle the leading '/' inside spiffs_find_objhdr_pgndx/spiffs_fobj_create, we can avoid the local variable and only change one place.

Reply:
one place? you mean three?

i can change it if you feel strongly.
but i guess it isn't the biggest issue for this PR.
the incompatibility with older nuttx is.

Reply:
> one place? you mean three?
> 

we need change at least spiffs_open, spiffs_unlink, spiffs_rename and spiffs_stat with the current patch. On the other hand, if we enhance spiffs_find_objhdr_pgndx to support the name without leading '\', all spiffs_xxx don't need change.
And these patches also don't need anymore:
fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
spiffs: Fix statfs f_namelen

> i can change it if you feel strongly.
> but i guess it isn't the biggest issue for this PR.
> the incompatibility with older nuttx is.

Yes, the padding added inside spiffs_pgobj_ndxheader_s is more critical than others. We need reference the offical spiffs to see how they define this structure.

Reply:
the structure with this PR should be compatible with:
* https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
* https://github.com/igrr/mkspiffs
* ESP-IDF spiffsgen.py

Reply:
> And these patches also don't need anymore:
> fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> spiffs: Fix statfs f_namelen

i'm not sure about this point.
do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.

Reply:
> the structure with this PR should be compatible with:
> 
> * https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
> * https://github.com/igrr/mkspiffs
> * ESP-IDF spiffsgen.py

If so, we need change the NuttX version to confirm the standard as soon as possible.

> > And these patches also don't need anymore:
> > fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> > spiffs: Fix statfs f_namelen
> 
> i'm not sure about this point.
> do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
> in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.

From your patch, '/' is always inserted before the relpath. If spiffs_find_objhdr_pgndx can handle relpath without the leading '/' correctly. Then we don't need reserve one space for '/'?

Reply:
done

Reply:
> > the structure with this PR should be compatible with:
> > 
> > * https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
> > * https://github.com/igrr/mkspiffs
> > * ESP-IDF spiffsgen.py
> 
> If so, we need change the NuttX version to confirm the standard as soon as possible.

it's what this PR does.

> 
> > > And these patches also don't need anymore:
> > > fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> > > spiffs: Fix statfs f_namelen
> > 
> > 
> > i'm not sure about this point.
> > do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
> > in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.
> 
> From your patch, '/' is always inserted before the relpath. If spiffs_find_objhdr_pgndx can handle relpath without the leading '/' correctly. Then we don't need reserve one space for '/'?

regardless of what nuttx do, those tools take a length including the leading slash.

in the latest push, i made it conditional for people who prefer compatibility with older nuttx.